infrastructure/src/main/java/org/wso2/testgrid/infrastructure/providers/aws/utils/AwsAMIMapper.java

```diff
@@ -0,0 +1,160 @@
+
+import com.amazonaws.auth.PropertiesFileCredentialsProvider;
+import com.amazonaws.services.ec2.AmazonEC2;
+import com.amazonaws.services.ec2.AmazonEC2ClientBuilder;
+import com.amazonaws.services.ec2.model.DescribeImagesRequest;
+import com.amazonaws.services.ec2.model.DescribeImagesResult;
+import com.amazonaws.services.ec2.model.Image;
+import com.amazonaws.services.ec2.model.Tag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.testgrid.common.exception.TestGridInfrastructureException;
+import org.wso2.testgrid.common.util.StringUtil;
+import org.wso2.testgrid.common.util.TestGridUtil;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * This class provides a mapper to find the matching AMI based on the infrastructure parameters of the test-plan.
+ */
+public class AwsAMIMapper {
+    private static final String AWS_REGION_NAME = "us-east-1";
+    private static final Logger logger = LoggerFactory.getLogger(AwsAMIMapper.class);
+
+    private final AmazonEC2 amazonEC2;
+
+    public AwsAMIMapper() throws IOException, TestGridInfrastructureException {
+        Path configFilePath = Paths.get(TestGridUtil.getTestGridHomePath(), "config.properties");
+        if (!configFilePath.toFile().exists()) {
+            throw new TestGridInfrastructureException(
+                    "Configuration file not found. Hence can not provide AWS credentials.");
+        }
+        amazonEC2 = AmazonEC2ClientBuilder.standard()
+                .withCredentials(new PropertiesFileCredentialsProvider(configFilePath.toString()))
+                .withRegion(AWS_REGION_NAME)
+                .build();
+    }
+
+    /**
+     * This method finds out the relevant AMI-id of the AMI which matches for the infra-parameters passed.
+     * @param infraParameters Infrastructure parameters (of the test-plan)
+     * @return AMI-id of the matching AMI
+     * @throws TestGridInfrastructureException When can not find a matching AMI
+     */
+    public String getAMI(Properties infraParameters) throws TestGridInfrastructureException {
+        List<Image> amiList = getAMIListFromAWS();
+        if (amiList.isEmpty()) {
+            throw new TestGridInfrastructureException("List of AMIs is empty. Hence can not find a matching AMI.");
+        }
+        //AMIs are containing tags only for the lookup-parameters.
+        Properties lookupParameters = filterLookupParameters(infraParameters);
+
+        Boolean isMissingLookupParams;  //False means one or more lookup parameters are missing.
+        String amiId = null;
+
+        //Check for an AMI which contains all (tags) lookupParameters.
+        for (Image ami : amiList) {
+            isMissingLookupParams = false;
+
+            for (String parameterName: lookupParameters.stringPropertyNames()) {
+                if (!checkIfParameterContains(
```

Comment:
I think we have three nested for-loops here which could be quite inefficient when the input data set increases. 
In checkIfParameterContains, you are looping the Tags for each parameter. May be we can optimize and remove the third for-loop. 

The variable names are a bit ambiguous. AWS SDK has a Parameter, and testgrid also have parameters. I'm not sure which is which by just looking at the variable names (such as parameterName, checkIfParameterContains).

Reply:
However, since we need to check "Per AMI" to ("Per Parameter" to "Per AMI Tag"), there is no way to avoid 3 nested loops (even if we use other method as contains, EOD it will be a loop). However based on the prediction of the number of AMIs, parameters and AMI tags, we can change the order of the nest.